src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationPerformanceMonitor.cs

```diff
@@ -2,89 +2,156 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
+using System.Linq;
 using UnityEngine;
 
 namespace Microsoft.MixedReality.SpectatorView
 {
-    internal enum StateSynchronizationPerformanceFeature : byte
+    public class StateSynchronizationPerformanceMonitor : Singleton<StateSynchronizationPerformanceMonitor>
     {
-        GameObjectComponentCheck = 0x0,
-        MaterialPropertyUpdate = 0x1,
-        MaterialPropertyBlockUpdate = 0x2,
-        Count = 0x3
-    }
-
-    internal class StateSynchronizationPerformanceMonitor : Singleton<StateSynchronizationPerformanceMonitor>
-    {
-        private const int PeriodsToAverageOver = 5;
-        private const int FeatureCount = (int)StateSynchronizationPerformanceFeature.Count;
-        private Stopwatch[] stopwatches;
-        private float[][] previousSpentTimes;
-        private float[][] previousActualTimes;
-        private int currentPeriod = 0;
+        private Dictionary<string, Dictionary<string, Stopwatch>> eventStopWatches = new Dictionary<string, Dictionary<string, Stopwatch>>();
+        private Dictionary<string, Dictionary<string, int>> eventCounts = new Dictionary<string, Dictionary<string, int>>();
 
         protected override void Awake()
         {
             base.Awake();
+        }
+
+        public IDisposable MeasureEventDuration(string componentName, string eventName)
+        {
+            if (!StateSynchronizationPerformanceParameters.EnablePerformanceReporting)
+            {
+                return null;
+            }
 
-            stopwatches = new Stopwatch[FeatureCount];
-            previousSpentTimes = new float[PeriodsToAverageOver][];
-            previousActualTimes = new float[PeriodsToAverageOver][];
-            for (int i = 0; i < FeatureCount; i++)
+            if (!eventStopWatches.TryGetValue(componentName, out var dictionary))
             {
-                stopwatches[i] = new Stopwatch();
+                dictionary = new Dictionary<string, Stopwatch>();
+                eventStopWatches.Add(componentName, dictionary);
             }
 
-            for (int i = 0; i < PeriodsToAverageOver; i++)
+            if (!dictionary.TryGetValue(eventName, out var stopwatch))
             {
-                previousSpentTimes[i] = new float[FeatureCount];
-                previousActualTimes[i] = new float[FeatureCount];
+                stopwatch = new Stopwatch();
+                dictionary.Add(eventName, stopwatch);
             }

```

Comment:
If we reuse stopwatch, this means we can't do event measurement re-entrancy?

Reply:
if we think supporting re-entrancy is important we can change the value definition of these dictionarys to a stop watch + flag. That going said, I'm not sure how we would choose to support re-stopping a stop watch in an intuitive manner once reentrancy occurs.

Reply:
We could build some sort of reference counting compared to running flag.

Reply:
I think for now well leave this as is, in the future we can invest more in reentrancy support as needed.